File: gesttravaux-java/src/main/java/com/immosync/gesttravaux/models/Devis.java
package com.immosync.gesttravaux.models;

import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * Entity representing a quote (devis) for a construction project.
 * A devis is associated with a prestation, entrepreneur, and chantier.
 *
 * @author GestTravaux Pro
 * @version 1.0
 */
@Entity
@Table(name = "DEVIS")
public class Devis {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "idDevis")
    private Integer idDevis;

    @Column(name = "prix")
    private Integer prix;

    @Column(name = "duree", length = 45)
    private String duree;

    @Column(name = "dateCreation")
    private LocalDateTime dateCreation;

    @Column(name = "statut", length = 20)
    private String statut;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "PRESTATIONS_idPrestation", nullable = false)
    private Prestation prestation;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "ENTREPRENEURS_idEntrepreneur", nullable = false)
    private Entrepreneur entrepreneur;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "CHANTIERS_idChantier", nullable = false)
    private Chantier chantier;

    /**
     * Default constructor required by JPA.
     */
    public Devis() {
        this.dateCreation = LocalDateTime.now();
        this.statut = "En attente";
    }

    /**
     * Constructor with essential fields.
     *
     * @param prix devis price
     * @param duree estimated duration
     * @param prestation associated prestation
     * @param entrepreneur assigned entrepreneur
     * @param chantier associated chantier
     */
    public Devis(Integer prix, String duree, Prestation prestation, Entrepreneur entrepreneur, Chantier chantier) {
        this.prix = prix;
        this.duree = duree;
        this.prestation = prestation;
        this.entrepreneur = entrepreneur;
        this.chantier = chantier;
        this.dateCreation = LocalDateTime.now();
        this.statut = "En attente";
    }

    // Getters and Setters

    /**
     * Gets the devis ID.
     *
     * @return the devis ID
     */
    public Integer getIdDevis() {
        return idDevis;
    }

    /**
     * Sets the devis ID.
     *
     * @param idDevis the devis ID
     */
    public void setIdDevis(Integer idDevis) {
        this.idDevis = idDevis;
    }

    /**
     * Gets the devis price.
     *
     * @return the price
     */
    public Integer getPrix() {
        return prix;
    }

    /**
     * Sets the devis price.
     *
     * @param prix the price
     */
    public void setPrix(Integer prix) {
        this.prix = prix;
    }

    /**
     * Gets the estimated duration.
     *
     * @return the duration
     */
    public String getDuree() {
        return duree;
    }

    /**
     * Sets the estimated duration.
     *
     * @param duree the duration
     */
    public void setDuree(String duree) {
        this.duree = duree;
    }

    /**
     * Gets the creation date.
     *
     * @return the creation date
     */
    public LocalDateTime getDateCreation() {
        return dateCreation;
    }

    /**
     * Sets the creation date.
     *
     * @param dateCreation the creation date
     */
    public void setDateCreation(LocalDateTime dateCreation) {
        this.dateCreation = dateCreation;
    }

    /**
     * Gets the devis status.
     *
     * @return the status
     */
    public String getStatut() {
        return statut;
    }

    /**
     * Sets the devis status.
     *
     * @param statut the status
     */
    public void setStatut(String statut) {
        this.statut = statut;
    }

    /**
     * Gets the associated prestation.
     *
     * @return the prestation
     */
    public Prestation getPrestation() {
        return prestation;
    }

    /**
     * Sets the associated prestation.
     *
     * @param prestation the prestation
     */
    public void setPrestation(Prestation prestation) {
        this.prestation = prestation;
    }

    /**
     * Gets the assigned entrepreneur.
     *
     * @return the entrepreneur
     */
    public Entrepreneur getEntrepreneur() {
        return entrepreneur;
    }

    /**
     * Sets the assigned entrepreneur.
     *
     * @param entrepreneur the entrepreneur
     */
    public void setEntrepreneur(Entrepreneur entrepreneur) {
        this.entrepreneur = entrepreneur;
    }

    /**
     * Gets the associated chantier.
     *
     * @return the chantier
     */
    public Chantier getChantier() {
        return chantier;
    }

    /**
     * Sets the associated chantier.
     *
     * @param chantier the chantier
     */
    public void setChantier(Chantier chantier) {
        this.chantier = chantier;
    }

    @Override
    public String toString() {
        return "Devis #" + idDevis + " - " + prix + " EUR - " + duree;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Devis)) return false;
        Devis devis = (Devis) o;
        return idDevis != null && idDevis.equals(devis.getIdDevis());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: gesttravaux-symfony/src/Entity/Devis.php
<?php

namespace App\Entity;

use App\Repository\DevisRepository;
use Doctrine\ORM\Mapping as ORM;

/**
 * Devis entity - Quotes/Estimates
 *
 * Represents contractor quotes for specific services on construction sites
 *
 * @author GestTravaux Pro
 */
#[ORM\Entity(repositoryClass: DevisRepository::class)]
#[ORM\Table(name: 'devis')]
class Devis
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idDevis')]
    private ?int $id = null;

    #[ORM\Column]
    private ?int $prix = null;

    #[ORM\Column(length: 45)]
    private ?string $duree = null;

    #[ORM\ManyToOne(targetEntity: Prestation::class, inversedBy: 'devis')]
    #[ORM\JoinColumn(name: 'prestations_idPrestation', referencedColumnName: 'idPrestation', nullable: false)]
    private ?Prestation $prestation = null;

    #[ORM\ManyToOne(targetEntity: Entrepreneur::class, inversedBy: 'devis')]
    #[ORM\JoinColumn(name: 'entrepreneurs_idEntrepreneur', referencedColumnName: 'idEntrepreneur', nullable: false)]
    private ?Entrepreneur $entrepreneur = null;

    #[ORM\ManyToOne(targetEntity: Chantier::class, inversedBy: 'devis')]
    #[ORM\JoinColumn(name: 'chantiers_idChantier', referencedColumnName: 'idChantier', nullable: false)]
    private ?Chantier $chantier = null;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getPrix(): ?int
    {
        return $this->prix;
    }

    public function setPrix(int $prix): static
    {
        $this->prix = $prix;

        return $this;
    }

    public function getDuree(): ?string
    {
        return $this->duree;
    }

    public function setDuree(string $duree): static
    {
        $this->duree = $duree;

        return $this;
    }

    public function getPrestation(): ?Prestation
    {
        return $this->prestation;
    }

    public function setPrestation(?Prestation $prestation): static
    {
        $this->prestation = $prestation;

        return $this;
    }

    public function getEntrepreneur(): ?Entrepreneur
    {
        return $this->entrepreneur;
    }

    public function setEntrepreneur(?Entrepreneur $entrepreneur): static
    {
        $this->entrepreneur = $entrepreneur;

        return $this;
    }

    public function getChantier(): ?Chantier
    {
        return $this->chantier;
    }

    public function setChantier(?Chantier $chantier): static
    {
        $this->chantier = $chantier;

        return $this;
    }

    /**
     * Get formatted price
     */
    public function getFormattedPrice(): string
    {
        return number_format($this->prix, 2, ',', ' ') . ' €';
    }

    public function __toString(): string
    {
        return 'Devis #' . $this->id . ' - ' . $this->getFormattedPrice();
    }
}
